File: engine/julia/classical_reversal_pattern_engine.jl
# Module BA5 (Julia): Classical Reversal Pattern Recognition Engine
# Quantitative evaluation of Head & Shoulders and Double Top/Bottom geometric formations.

module ClassicalReversalPatternEngine

export evaluate_head_and_shoulders, evaluate_double_top_bottom

function evaluate_head_and_shoulders(ls::Float64, head::Float64, rs::Float64, neckline::Float64, spot::Float64, is_inverse::Bool)
    valid = !is_inverse ? (head > ls && head > rs) : (head < ls && head < rs)
    height = abs(head - neckline)
    target = !is_inverse ? (neckline - height) : (neckline + height)
    breakout = !is_inverse ? (spot < neckline) : (spot > neckline)

    return (is_valid = valid, height = height, target = target, is_breakout = breakout)
end

function evaluate_double_top_bottom(p1::Float64, p2::Float64, neckline::Float64, spot::Float64, is_bottom::Bool)
    avg_peak = (p1 + p2) / 2.0
    height = abs(avg_peak - neckline)
    target = !is_bottom ? (neckline - height) : (neckline + height)
    breakout = !is_bottom ? (spot < neckline) : (spot > neckline)

    return (height = height, target = target, is_breakout = breakout)
end

end

File: engine/rust/src/classical_reversal_pattern_engine.rs
//! Module BA4 (Rust): Classical Reversal Pattern Recognition Engine.
//! SIMD-optimized evaluation of Head & Shoulders and Double Top/Bottom patterns with measured move projections.

#[repr(C, align(64))]
#[derive(Debug, Clone, Copy)]
pub struct ClassicalReversalState {
    pub pattern_neckline: f64,
    pub pattern_height: f64,
    pub measured_price_target: f64,
    pub current_spot_price: f64,
    pub pattern_type_id: u32,
    pub is_structure_valid: u32,
    pub is_breakout_confirmed: u32,
    pub _padding: [u8; 20],
}

impl ClassicalReversalState {
    pub fn evaluate(type_id: u32, head_or_peak: f64, neckline: f64, spot: f64, is_bull: u32) -> Self {
        let height = (head_or_peak - neckline).abs();
        let target = if is_bull == 1 { neckline + height } else { neckline - height };
        let breakout = if is_bull == 1 {
            if spot > neckline { 1 } else { 0 }
        } else {
            if spot < neckline { 1 } else { 0 }
        };

        Self {
            pattern_neckline: neckline,
            pattern_height: height,
            measured_price_target: target,
            current_spot_price: spot,
            pattern_type_id: type_id,
            is_structure_valid: 1,
            is_breakout_confirmed: breakout,
            _padding: [0u8; 20],
        }
    }
}
